pkg/pilot/genericpilot/controller_test.go

```diff
@@ -0,0 +1,268 @@
+			kubeObjects:            []runtime.Object{newDummyPilot("name", "namespace")},
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
+		},
+		{
+			name:                   "should not Start a subprocess when syncing a non-this pilot",
+			key:                    "namespace/not-this-pilot",
+			err:                    false,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/not-this-pilot"},
+			cachedThisPilot:        newDummyPilot("name", "namespace"),
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("not-this-pilot", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("not-this-pilot", "namespace")},
+		},
+		{
+			name:                   "failed genericpilot consumer sync function should fail sync",
+			key:                    "namespace/name",
+			err:                    true,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/name"},
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("name", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("name", "namespace")},
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
+		},
+		{
+			name:               "missing cachedThisPilot and empty clientset should fail Sync()",
+			key:                "namespace/name",
+			err:                true,
+			thisPilotName:      "name",
+			thisPilotNamespace: "namespace",
+		},
+		{
+			name: "should reuse cached 'this' pilot resource, and should fail to update status",
+			key:  "namespace/name",
+			err:  true,
+			syncFuncExpectedPilot:  newDummyPilot("name", "namespace"),
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/name"},
+			cachedThisPilot:        newDummyPilot("name", "namespace"),
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
+		},
+		{
+			name: "should sync not-this-pilot and should not update the pilots status",
+			key:  "namespace/not-this-pilot",
+			err:  false,
+			syncFuncExpectedPilot:  newDummyPilot("not-this-pilot", "namespace"),
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/not-this-pilot"},
+			cachedThisPilot:        newDummyPilot("name", "namespace"),
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("not-this-pilot", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("not-this-pilot", "namespace")},
+			process:                newDefaultProcessAdapter(),
```

Comment:
Again, I don't understand this test description. `should sync not-this-pilot and should not update the pilots status`.
Doesn't this show that the pilot ignores foreign / peer pilot resources.
But in that case why is the foreign / peer pilot among the `expectedScheduledItems`? 
To me, that implies that the Pilot controller has looked at the pilot, found it to be foreign and then put it back on the queue, rather than discarding it.

Reply:
So 'not-this-pilot' in this case *is* a peer/member of the cluster, but it is not *this* pilot.

This means that it *should* be synced (right now we don't support leader election, but once we do this becomes a lot more important). But it *shouldn't* update the Status block of that Pilot as a Pilot is only responsible for updating it's *own* status block.

Reply:
It's then expected to be scheduled again so that things 'keep moving'. Perhaps in future we won't reschedule peers every 10s, but for now it saves queues getting blocked because there's been no change to any resource (so nothing happens)